Add item requests

Aleks312T opened the pull request.

Aleks312T committed:
feat: реализованы ItemRequestDto и его вариации
feat: добавлены заготовки для сервисов запросов

Aleks312T committed:
feat: реализованы несколько методов в ItemRequestServiceImpl
feat: создан ItemRequestMapper

Aleks312T committed:
feat: добавлены методы в ItemRequestController
fix: исправлены мелкие ошибки (725 тестов из 780)

Aleks312T committed:
feat: закончены методы в ItemRequestServiceImpl
fix: исправлены мелкие ошибки (743 тестов из 780)

Aleks312T committed:
feat: добавлены корректировки для Item относительно ItemRequest
fix: исправлены мелкие ошибки (772 тестов из 780)

Aleks312T committed:
feat: добавлены корректировки для Booking относительно ItemRequest
fix: исправлены тесты (780 тестов из 780)

Aleks312T committed:
feat: добавлены тесты UserControllerTest и ItemControllerTest

Aleks312T committed:
feat: добавлены тесты BookingControllerTest и ItemRequestControllerTest
fix: исправлена ошибка, связанная с проверкой входных данных
delete: удалены неиспользуемые исключения

Aleks312T committed:
feat: начата работа над тестами для UserService

Aleks312T committed:
feat: доделаны UserServiceImplTest

Aleks312T committed:
feat: сделаны тесты для ItemService

Aleks312T committed:
feat: сделаны тесты для BookingService

Aleks312T committed:
feat: сделаны тесты для ItemRequestController

Aleks312T committed:
style: чистка кода

Aleks312T committed:
style: убраны ненужные зависимости

Aleks312T committed:
feat: улучшение покрытия кода

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,14 +1,18 @@
+import javax.validation.constraints.Min;
 import java.util.List;
 
+@Valid
```
Valid
Вопрос: а зачем здесь эта аннотация? 

Различия в аннотациях @Valid и @Validated в Spring  
https://for-each.dev/lessons/b/-spring-valid-vs-validated

Aleks312T replied:
Понял, спасибо. Видимо от нервов добавил, пока пытался исправить ошибку с проверкой входных данных.

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -104,37 +106,45 @@ public BookingUserDto confirm(Long bookingId, long userId, boolean approved) {
 
     @Transactional
```
readonly
Если метод не меняет данные в базе, то лучше добавлять этот параметр.

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -205,7 +212,7 @@ public void checkBooking(Booking booking) {
                 || booking.getEnd().isBefore(booking.getStart())
                 || booking.getEnd().isEqual(booking.getStart())) {
-            throw new IncorrectParameterException("Ошибка во времени бронирования");
+            throw new ValidationException("Ошибка во времени бронирования");
```
Всё-таки лучше добавлять конкретики в такого рода сообщения, чтобы было легче и быстрее разобраться в чём ошибка.

avfyodorov commented on src/main/java/ru/practicum/shareit/request/dto/ItemRequestFullDto.java:
```diff
@@ -0,0 +1,30 @@
+    @NotNull
+    private LocalDateTime created;
+
+    List<ItemDto> items;
```
И это поле тоже лучше сделать приватным, не нужно давать доступ извне.

Aleks312T replied:
Исправил, но меня смущает слово "тоже". Я вроде прошелся по всем dto и больше открытых полей не нашел.

avfyodorov commented on src/main/java/ru/practicum/shareit/request/service/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,125 @@
+    @Transactional
+    public ItemRequestFullDto create(Long userId, ItemRequestDtoInput requestDtoInput) {
+        log.debug("Вызов метода create с userId = {}", userId);
+        //Не использую функцию, чтобы не обращаться к БД два раза
```
Понятно.

avfyodorov reviewed (commented):
Добрый день, Александр! 
Отличная работа, от меня только пара небольших уточнений, а так всё в порядке.

Aleks312T committed:
feat: исправлены замечания
feat: написаны еще несколько тестов
fix: изменены зависимости для Transactional

avfyodorov reviewed (commented):
Добрый вечер, Александр! 
Всё в порядке, Замечаний нет. 
Работа принята. 

“Тоже” -это в том смысле, что предыдущие два поля в классе уже приватные. 
Неудачно выразиться, пардон.  🙂
